detox.1
.\" 
.\" Copyright (c) 2004-2017, Doug Harple.  All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\" 
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" 3. Neither the name of author nor the names of its contributors may be
.\"    used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.Dd August 3, 2004
.Dt DETOX 1
.Os
.Sh NAME
.Nm detox
.Nd clean up filenames
.Sh SYNOPSIS
.Nm
.Op Fl hnLrv
.Op Fl s sequence
.Op Fl f configfile
.Op Fl -dry-run
.Op Fl -special
.Ar
.Sh DESCRIPTION
The
.Nm
utility renames files to make them easier to work with.  It removes
spaces and other such annoyances.  It'll also translate or cleanup
Latin-1 (ISO 8859-1) characters encoded in 8-bit ASCII, Unicode
characters encoded in UTF-8, and CGI escaped characters.
.Ss Sequences
.Nm
is driven by a configurable series of filters, called a sequence.
Sequences are covered in more detail in
.Xr detoxrc 5
and are discoverable with the
.Fl L
option.  Some examples of default sequences are
.Cm iso8859_1
and
.Cm utf_8 .
.Ss Options
The main options:
.Bl -tag -width Fl
.It Fl f Ar configfile
Use 
.Ar configfile 
instead of the default configuration files for loading translation
sequences.  No other config file will be parsed.
.It Fl h -help
Display helpful information.
.It Fl L
List the currently available sequences.  When paired with
.Fl v
this option shows what filters are used in each sequence and any
properties applied to the filters.
.It Fl n -dry-run
Doesn't actually change anything.  This implies the
.Fl v
option.
.It Fl r
Recurse into subdirectories.
.It Fl s Ar sequence
Use
.Ar sequence 
instead of default.
.It Fl -special
Works on special files (including links).  Normally 
.Nm
ignores these files.
.It Fl v
Be verbose about which files are being renamed.
.It Fl V
Show the current version of
.Nm .
.El
.Ss Deprecated Options
Deprecated Options are options that were available in earlier versions
of
.Nm
but have lost their meaning and are being phased out.
.Bl -tag -width Fl
.It Fl -remove-trailing
Removes _ and - after .'s in filenames.  This was first provided in
the 0.9 series of 
.Nm .
After the introduction of sequences, it lost its meaning, as you could
now determine the properties of wipeup through a particular sequence's
configuration.  It presently forces all instances of the wipeup filter
to use remove trailing, regardless of what's actually in the config
files.
.El
.Sh FILES
.Bl -tag -width iso8859_1.tbl -compact
.It Pa detoxrc
The system-wide detoxrc file.
.It Pa ~/.detoxrc
A user's personal detoxrc.  Normally it extends the system-wide
detoxrc, unless
.Fl f
has been specified, in which case, it is ignored.
.It Pa iso8859_1.tbl
The default ISO 8859-1 translation table.
.It Pa unicode.tbl
The default Unicode (UTF-8) translation table.
.El
.Sh EXAMPLES
.Bl -tag -width Fl
.It Nm Fl s Ar iso8859_1 Fl r Fl v Fl n Pa /tmp/new_files
Will run the sequence
.Ar iso8859_1
recursively, listing any changes, without changing anything, on the
files of
.Pa /tmp/new_files .
.It Nm Fl c Ar my_detoxrc Fl L Fl v
Will list the sequences within
.Ar my_detoxrc ,
showing their filters and options.
.El
.Sh SEE ALSO
.Xr detoxrc 5 ,
.Xr detox.tbl 5 .
.Sh HISTORY
.Nm 
was originally designed to clean up files that I had received from
friends which had been created using other operating systems.  It's
trivial to create a filename with spaces, parenthesis, brackets, and
ampersands under some operating systems.  These have special meaning
within FreeBSD and Linux, and cause problems when you go to access
them.  I created
.Nm
to clean up these files.
.Sh AUTHORS
.Nm
was written by
.An "Doug Harple" .
.Sh BUGS
If, after the translation of a filename is finished, a file already
exists with that same name,
.Nm
will not rename the file.  This could cause a problem with the
.Cm max_length
filter, if it was imperative that the files be cut down to a certain
length.
.Pp
Long options don't work under Solaris or Darwin.
.Pp
An error in the config file will cause a segfault as it's going to
print the offending word within the config file.


CHANGES
Version 1.2.0:
	Modified the safe filter to use a translation table.
	Modified the safe filter fallback (previous functionality) to operate
		without any special behavior.  The wipeup filter now picks up
		where the safe filter left off.
	Fixed the default permissions on install (files are 644 now).
	Updated libpopt support to work on Linux under the PowerPC platform
		(chars are unsigned by default).
	Included the generated lex and yacc files in the default package.
	Added additional logic to allow files on case insensitive filesystems
		to have their case changed.
	Added the ability to set locale specific translations in the
		translation tables.
	Added German specific translations to the translation tables.
	Added the ability to ignore specific files.
		[sourceforge.net tracker #1253826]
	Fixed a bug where directories specified on the command line wouldn't
		get translated. [sourceforge.net tracker #1213623]
	Added support for translating large files.
		[sourceforge.net tracker #1509493]
	Added inline-detox for stream based detoxification.

Version 1.1.1:
	Modified Makefile to support parallel builds.
	Added ${DESTDIR} to install paths, for Gentoo package builds.
	Modified the install script to not overwrite existing configuration
		files or translation tables.
	Modified the install script to install the config file and translation
		tables as ".sample" as well as the working version, for all
		users, but in particular, to make patching the Makefile easier
		for the FreeBSD port.

Version 1.1.0:
	Added lowercase filter.
	Added libpopt support to facilitate long options on Darwin or
		Solaris.
	Fixed some compiler gripes with lex/yacc.
	Replaced the hardcoded -ll in Makefile.in with @LEXLIB@.

Version 1.0.0:
	Added a new filter for translating UTF-8 encoded Unicode characters.
	Added handling of configuration files for controlling what sequence
		filters are run in.
	Added handling of loadable translation tables, so the user can control
		how the ISO 8859-1 and Unicode filters operate.
	Added a new filter for trimming based on the max length.
	Added command line options: 
		-f	set config file
		-L	list sequences
		-n	the same as --dry-run
		-s	set sequence
	Added handling for an environmental variable DETOX_SEQUENCE, which
		sets the default sequence name.
	Translation of some Icelandic characters has changed.  0xd0, 0xde,
		0xf0, 0xfe, the Icelandic characters for "Eth" and "Thorn" have
		been changed from "D", "Y", "o", "y" to "TH" and "th".
	Fixed translation of 0xfc (u), 0xfd (y) and 0xff (y).
	Added .depend generation to the Makefile.
	Created more man pages (detoxrc.5 and detox.tbl.5).

Version 0.9.1:
	Added -d flag to install 
	Broke installation out into a script to handle differences between
		Solaris and BSD/Linux.
	Added function check for getopt_long.

Version 0.9.0:
	Initial release


detox.c
/*
 * Copyright (c) 2004-2017, Doug Harple.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of author nor the names of its contributors may be
 *    used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

#include "detox.h"
#include "clean_string.h"
#include "detox_path.h"

#ifndef INLINE_MODE
#include "file.h"
#else
#include "file-inline.h"
#endif

#include "config_file.h"
#include "config_file_spoof.h"
#include "config_file_dump.h"
#include "parse_table.h"

#ifndef HAVE_LIBPOPT
#include "parse_options_getopt.h"
#else
#include "parse_options_popt.h"
#endif

#define MAX_PATH_LEN 256

int main(int argc, char **argv)
{
	struct stat stat_info;
	int err;

	struct detox_parse_results *parse_results = NULL;
	struct detox_sequence_list *list_work = NULL;
	struct detox_sequence_entry *which_sequence = NULL;
	struct detox_sequence_entry *work = NULL;
	struct detox_options *main_options;

	char *check_config_file = NULL;
	char *file_work = NULL;
	char **file_walk;

#ifndef HAVE_LIBPOPT
	main_options = parse_options_getopt(argc, argv);
#else
	main_options = parse_options_popt(argc, argv);
#endif

	if (main_options == NULL) {
		fprintf(stderr, "detox: an error occurred while parsing command line arguments\n");
		exit(EXIT_FAILURE);
	}

	if (main_options->check_config_file) {
		check_config_file = strdup(main_options->check_config_file);
	}

	if (check_config_file != NULL) {
		parse_results = parse_config_file(check_config_file, NULL, main_options);
		if (parse_results == NULL) {
			fprintf(stderr, "detox: unable to open: %s\n", check_config_file);
			exit(EXIT_FAILURE);
		}
	}
	else {
		check_config_file = malloc(MAX_PATH_LEN);
		if (check_config_file == NULL) {
			fprintf(stderr, "out of memory: %s\n", strerror(errno));
			exit(EXIT_FAILURE);
		}

		err = snprintf(check_config_file, MAX_PATH_LEN, "%s/detoxrc", default_etc_dir);
		if (err < MAX_PATH_LEN)
			parse_results = parse_config_file(check_config_file, NULL, main_options);

		if (parse_results == NULL) {
			parse_results = parse_config_file("/etc/detoxrc", NULL, main_options);
		}

		if (parse_results == NULL) {
			parse_results = parse_config_file("/usr/local/etc/detoxrc", NULL, main_options);
		}

		file_work = getenv("HOME");
		if (file_work != NULL) {
			err = snprintf(check_config_file, MAX_PATH_LEN, "%s/.detoxrc", file_work);
			if (err < MAX_PATH_LEN)
				parse_results = parse_config_file(check_config_file, parse_results, main_options);

			file_work = NULL;
		}

		if (parse_results == NULL) {
			parse_results = spoof_config_file(main_options);
		}

		free(check_config_file);
	}

	if (parse_results == NULL) {
		fprintf(stderr, "detox: no config file to work with\n");
		exit(EXIT_FAILURE);
	}

	/*
	 * Store the files_to_ignore array in the main_options struct for use in
	 * parse_dir/file/special
	 */

	main_options->files_to_ignore = parse_results->files_to_ignore;

	/*
	 * Determine which sequence to use
	 */

	which_sequence = NULL;

	list_work = parse_results->sequences;

	while (list_work != NULL) {
		if (strcmp(list_work->name, (main_options->sequence_name == NULL) ? "default" : main_options->sequence_name) == 0) {
			which_sequence = list_work->head;
			break;
		}

		list_work = list_work->next;
	}

	/*
	 * If no sequence was found, and the user didn't specify a sequence
	 * to use, just use the first sequence.
	 */

	if (which_sequence == NULL && main_options->sequence_name == NULL) {
		if (parse_results->sequences != NULL) {
			which_sequence = parse_results->sequences->head;
		}
	}

	main_options->sequence_to_use = which_sequence;

	/*
	 * List sequences
	 */
	if (main_options->list_sequences) {
		dump_config_file(parse_results, main_options);
		exit(EXIT_SUCCESS);
	}

	/*
	 * Fail if no sequence is available
	 */
	if (main_options->sequence_to_use == NULL) {
		/*
		 * XXX - Explain this better
		 */
		fprintf(stderr, "detox: no sequence to work with\n");
		exit(EXIT_FAILURE);
	}

	/*
	 * Check translation tables
	 */

	work = main_options->sequence_to_use;
	while (work != NULL) {
		char *check_filename = NULL;
		int do_search = 0;

		struct translation_table *table = NULL;
		struct clean_string_options *opts;

		if (work->cleaner == &clean_iso8859_1) {
			if (work->options != NULL) {
				opts = work->options;
				if (opts->filename != NULL) {
					check_filename = opts->filename;
				}
			}

			if (!check_filename) {
				check_filename = "iso8859_1.tbl";
				do_search = 1;
			}
		}
		else if (work->cleaner == &clean_utf_8) {
			if (work->options != NULL) {
				opts = work->options;
				if (opts->filename != NULL) {
					check_filename = opts->filename;
				}
			}

			if (!check_filename) {
				check_filename = "unicode.tbl";
				do_search = 1;
			}
		}
		else if (work->cleaner == &clean_safe) {
			if (work->options != NULL) {
				opts = work->options;
				if (opts->filename != NULL) {
					check_filename = opts->filename;
				}
			}

			if (!check_filename) {
				check_filename = "safe.tbl";
				do_search = 1;
			}
		}

		if (check_filename || do_search) {

			table = NULL;

			if (do_search) {
				check_config_file = malloc(MAX_PATH_LEN);
				if (check_config_file == NULL) {
					fprintf(stderr, "out of memory: %s\n", strerror(errno));
					exit(EXIT_FAILURE);
				}

				err = snprintf(check_config_file, MAX_PATH_LEN, "%s/%s", default_data_dir, check_filename);
				if (err < MAX_PATH_LEN)
					table = parse_table(check_config_file);

				if (table == NULL) {
					err = snprintf(check_config_file, MAX_PATH_LEN, "/usr/share/detox/%s", check_filename);
					if (err < MAX_PATH_LEN)
						table = parse_table(check_config_file);
				}

				if (table == NULL) {
					err = snprintf(check_config_file, MAX_PATH_LEN, "/usr/local/share/detox/%s", check_filename);
					if (err < MAX_PATH_LEN)
						table = parse_table(check_config_file);
				}

				if (table == NULL) {

					/*
					 * Fall back to the non-file based
					 * cleaner
					 */
					if (work->cleaner == &clean_iso8859_1) {
						work->cleaner = &clean_iso8859_1_basic;
					}
					else if (work->cleaner == &clean_utf_8) {
						work->cleaner = &clean_utf_8_basic;
					}
					else if (work->cleaner == &clean_safe) {
						work->cleaner = &clean_safe_basic;
					}
					else {
						fprintf(stderr, "detox: unable to locate translation table or fall back\n");
						exit(EXIT_FAILURE);
					}
				}
				else {

					/*
					 * Allocate an options
					 */
					opts = malloc(sizeof(struct clean_string_options));
					if (opts == NULL) {
						fprintf(stderr, "out of memory: %s\n", strerror(errno));
						exit(EXIT_FAILURE);
					}
					memset(opts, 0, sizeof(struct clean_string_options));

					opts->translation_table = table;
					work->options = opts;
				}

				free(check_config_file);
			}
			else {
				table = parse_table(check_filename);
				if (table == NULL) {
					fprintf(stderr, "detox: unable to parse file: %s\n", check_filename);
					exit(EXIT_FAILURE);
				}

				opts = work->options;
				opts->translation_table = table;
			}
		}


		work = work->next;
	}

	/*
	 * Do some actual work
	 */

#ifndef INLINE_MODE
	file_walk = main_options->files;
	while (*file_walk) {
		if (main_options->verbose) {
			printf("Scanning: %s\n", *file_walk);
		}

		err = lstat(*file_walk, &stat_info);
		if (err == -1) {
			fprintf(stderr, "%s: %s\n", *file_walk, strerror(errno));
		}
		else {
			if (S_ISDIR(stat_info.st_mode)) {
				file_work = parse_file(*file_walk, main_options);
				parse_dir(file_work, main_options);
				free(file_work);
			}
			else if (S_ISREG(stat_info.st_mode)) {
				parse_file(*file_walk, main_options);
			}
			else if (main_options->special) {
				parse_special(*file_walk, main_options);
			}
		}

		file_walk++;
	}
#else

	if (main_options->files[0] != NULL) {
		file_walk = main_options->files;
		while (*file_walk) {
			err = lstat(*file_walk, &stat_info);
			if (err == -1) {
				fprintf(stderr, "%s: %s\n", *file_walk, strerror(errno));
			}
			else {
				if (S_ISDIR(stat_info.st_mode)) {
					fprintf(stderr, "%s: is a directory\n", *file_walk);
				}
				else {
					parse_inline(*file_walk, main_options);
				}
			}

			file_walk++;
		}
	}
	else {
		parse_inline(NULL, main_options);
	}

#endif

	return 0;
}
